design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,160 @@
+
+However, on nodes with [hyperthreading](https://en.wikipedia.org/wiki/Hyper-threading) (SMT) enabled, this causes an error, described on [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103).
+This design proposal is meant to fix this issue.
+
+Considering that hyperthreading (SMT) is enabled by default in many [Intel](https://www.intel.com/content/www/us/en/architecture-and-technology/hyper-threading/hyper-threading-technology.html) processors - it means that DPDK and Realtime users may be impacted.
+
+## Motivation
+Running High performance VMs with an even number of CPUs with the conditions below.
+
+### The problem
+In order to replicate this issue, the following conditions must be met:
+* The node has hyperthreading (SMT) enabled.
+* Kubelet’s CPUManager is set to static - full-pcpus-only.
+* Kubevirt CR's defaultRuntimeClass is set to a high performance runtimeClass.
+* VM is set with `dedicatedCPUPlacement` and `IsolateEmulatorThread` enabled.
+* VM requests an even number of CPUs.
+
+When enabling `IsolateEmulatorThread` on a VM with even CPU count, Kubevirt is requesting one extra CPU for the virt-launcher pod. This is an internal change that is not exposed to the user.
+After the pod is scheduled, the now odd number of CPUs will collide with the CPU-Manager policy of using only full physical cores, and cause the kubelet to [reject](https://github.com/kubernetes/kubernetes/pull/101432) this pod.
+For example, let’s consider this VM spec:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        cpu:
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+          sockets: 1
+          cores: 4
+          threads: 2
+```
+kubelet will emit following event:
+```
+SMT Alignment Error: requested 9 cpus not multiple cpus per core = 2
+```
+The following picture helps show how this is a problem
+<img src="smt-slignment-error-diagram.png">
+
+## Goals
+Fix [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103) on the Kubevirt scope.
+
+## Non  Goals
+Fix the actual scheduling of the virt-launcher to a node that will reject it.
+
+## Definition of Users
+This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.
+
+## User Stories
+A user should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.
+
+## Repos
+- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Bug fix design proposal
+The bug fix proposal introduces an alpha annotation `kubevirt.io/full-pcpus-only: true`:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    metadata:
+      annotations:
+        kubevirt.io/full-pcpus-only: true
```

Comment:
Please consider having an annotation without a value.

Reply:
It fits the convention of `isolateEmulatorThread: true` (though this one is an API)
So I can go either way. 
Any objections to setting the annotation to `kubevirt.io/full-pcpus-only:`?

Reply:
Perhaps, it would be better to match the CPU manager's policy wording and change this to something like, `kubevirt.io/CPUManagerPolicyBetaOptions:full-pcpus-only`
This way it will be clear where this is coming from and may accommodate other policies there might (in the future - hopefully we won't get to this)

Reply:
DONE